controllers/actions.github.com/autoscalingrunnerset_controller_test.go

```diff
@@ -1465,3 +1551,76 @@ var _ = Describe("Test external permissions cleanup", func() {
 		).Should(BeTrue(), "Expected role to be cleaned up")
 	})
 })
+
+var _ = Describe("Version mismatch cleanup", func() {
+	It("Should delete autoscaling runner set on a version mismatch", func() {
+		ctx := context.Background()
+		autoscalingNS, mgr := createNamespace(GinkgoT(), k8sClient)
+
+		configSecret := createDefaultSecret(GinkgoT(), k8sClient, autoscalingNS.Name)
+
+		controller := &AutoscalingRunnerSetReconciler{
+			Client:                             mgr.GetClient(),
+			Scheme:                             mgr.GetScheme(),
+			Log:                                logf.Log,
+			ControllerNamespace:                autoscalingNS.Name,
+			DefaultRunnerScaleSetListenerImage: "ghcr.io/actions/arc",
+			ActionsClient:                      fake.NewMultiClient(),
+		}
+		err := controller.SetupWithManager(mgr)
+		Expect(err).NotTo(HaveOccurred(), "failed to setup controller")
+
+		originalVersion := build.Version
+		defer func() {
+			build.Version = originalVersion
+		}()
+		build.Version = "0.2.0"
+
+		min := 1
+		max := 10
+		autoscalingRunnerSet := &v1alpha1.AutoscalingRunnerSet{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      "test-asrs",
+				Namespace: autoscalingNS.Name,
+				Labels: map[string]string{
+					"app.kubernetes.io/name":    "gha-runner-scale-set",
+					"app.kubernetes.io/version": "0.1.0",
+				},
+				Annotations: map[string]string{
+					AnnotationKeyKubernetesModeRoleBindingName:    "kube-mode-role-binding",
+					AnnotationKeyKubernetesModeRoleName:           "kube-mode-role",
+					AnnotationKeyKubernetesModeServiceAccountName: "kube-mode-service-account",
+				},
+			},
+			Spec: v1alpha1.AutoscalingRunnerSetSpec{
+				GitHubConfigUrl:    "https://github.com/owner/repo",
+				GitHubConfigSecret: configSecret.Name,
+				MaxRunners:         &max,
+				MinRunners:         &min,
+				RunnerGroup:        "testgroup",
+				Template: corev1.PodTemplateSpec{
+					Spec: corev1.PodSpec{
+						Containers: []corev1.Container{
+							{
+								Name:  "runner",
+								Image: "ghcr.io/actions/runner",
+							},
+						},
+					},
+				},
+			},
+		}
+
+		// create autoscaling runner set before starting a manager
+		err = k8sClient.Create(ctx, autoscalingRunnerSet)
+		Expect(err).NotTo(HaveOccurred())
+
+		startManagers(GinkgoT(), mgr)
+
+		Eventually(func() bool {
+			ars := new(v1alpha1.AutoscalingRunnerSet)
+			err := k8sClient.Get(ctx, types.NamespacedName{Namespace: autoscalingRunnerSet.Namespace, Name: autoscalingRunnerSet.Name}, ars)
+			return errors.IsNotFound(err)
+		}).Should(BeTrue())
+	})
```

Comment:
I'm not really sure how we're testing that the deletion / creation actually happened, could you help me understand this part please?

Reply:
Sure, so once we issue a `k8sClient.Create`, it is going to create a resource in etcd.
Then our controller is triggered, we fetch the autoscaling runner set and the controller should check the version.
The resource is intentionally made with the version mismatch, so eventually, the controller needs to delete this resource as soon as it can.
Every other test have resource version matching the `build.Version`, so when we run tests, happy path is covered.
This test tries make sure that we are actually going to delete this resource because of the version mismatch. Eventually is there because we can't know the exact time this is going to be called, and with finalizer checks, it can take some time.
And since we are configuring the autoscaling runner set with correct parameters, but incorrect version, the deletion should happen without anyone issuing a `DELETE` request